services/wfm/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteService.java

```diff
@@ -263,14 +289,25 @@ private boolean filterForwardPrimaryPath(FlowPath path) {
     public Map<Flow, Set<PathId>> getInactiveFlowsForRerouting() {
         log.info("Get inactive flows");
         return flowRepository.findDownFlows().stream()
-                .collect(Collectors.toMap(Function.identity(),
+                .collect(toMap(Function.identity(),
                         flow -> flow.getPaths().stream()
                         .filter(path -> FlowPathStatus.INACTIVE.equals(path.getStatus()))
                         .map(FlowPath::getPathId)
                         .collect(Collectors.toSet()))
                 );
     }
 
+    /**
+     * Returns affected inactive flow set for rerouting.
+     */
+    public Set<Flow> getAffectedInactiveFlowsForRerouting(SwitchId switchId) {
+        log.info("Get affected inactive flows for switch {}", switchId);
+        return flowPathRepository.findBySegmentSwitch(switchId).stream()
+                .map(FlowPath::getFlow)
+                .filter(f -> f.getStatus() == FlowStatus.DOWN || f.getStatus() == FlowStatus.DEGRADED)
```

Comment:
it can be better. You can optimize it by getting only Flow Paths of Down and Degraded flows by Neo4j.

To do it you need to add new method into flowPathRepository. Something like

```

    @Override
    public Collection<FlowPath> findInactiveBySegmentSwitch(SwitchId switchId) {
        Map<String, Object> parameters = ImmutableMap.of(
                "switch_id", switchIdConverter.toGraphProperty(switchId));

        Set<String> pathIds = new HashSet<>();
        queryForStrings("MATCH (ps_src:switch)-[:source]-(ps:path_segment)-[:destination]-(ps_dst:switch)" +
                "WHERE ps_src.name = $switch_id OR ps_dst.name = $switch_id" +
                "MATCH (f:flow)-[:owns]-(fp:flow_path)-[:owns]-(ps)" +
                "WHERE f.status = 'down' or f.status = 'degraded'" +
                "RETURN fp.path_id as path_id", parameters, "path_id").forEach(pathIds::add);

        if (pathIds.isEmpty()) {
            return emptyList();
        }

        Filter pathIdsFilter = new Filter(PATH_ID_PROPERTY_NAME, new InOperatorWithNoConverterComparison(pathIds));

        return loadAll(pathIdsFilter);
    }
```

Reply:
Reworked